### PR TITLE
enhancement(sources): Tls for splunk and logplex

### LINKED ISSUE #1904: Add support for TLS to the logplex and splunk_hec sources
These two sources are the last ones not covered by previous efforts (#1892 and #1893) as they are based on the `warp` crate for HTTP support. I have done the naïve support but it pulls in the `rustls` crate among other things, which is something we have decided against doing. @LucioFranco indicated he was willing to help out with this process, as he was familiar with the issues and knew it would not be entirely straightforward to complete.

Closes #1553

### PR DESCRIPTION
**Pull request by LucioFranco**
We now maintain a fork of warp for timber's usage but this should be removed once we update to warp 0.2 which has finally exposed things to allow us to customize the transport.

This also implements an abstract `MaybeTlsStream` type which abstracts around a tls enabled io type and a plain one without the end user having to know. 

Closes #1904 

Signed-off-by: Lucio Franco <luciofranco14@gmail.com>

### CODE REVIEW

**Comment by LucioFranco** [comment]
I will add docs shortly but wanted to get the code up for review.

File: src/tls.rs [hunk 1]
```diff
@@ -256,6 +261,137 @@ fn open_read(filename: &Path, note: &'static str) -> crate::Result<Vec<u8>> {
...
+
+                        self.state = MaybeTlsIncomingState::Accepting(fut);
+                        continue;
+                    } else {
+                        return Ok(Async::Ready(Some(MaybeTlsStream::Raw(stream))));
+                    }
+                }
+
+                MaybeTlsIncomingState::Accepting(fut) => {
+                    let stream = try_ready!(fut.poll().map_err(Box::new));
```
Comment (bruceg) [inline src/tls.rs]: Does this error provide enough context to be able to parse the error, or would it be helpful to wrap it with a (snafu) context?

File: src/tls.rs [hunk 2]
```diff
@@ -256,6 +261,137 @@ fn open_read(filename: &Path, note: &'static str) -> crate::Result<Vec<u8>> {
...
+        match self {
+            MaybeTlsStream::Tls(s) => s.read(buf),
+            MaybeTlsStream::Raw(s) => s.read(buf),
+        }
+    }
+}
+
+impl<S: AsyncRead + AsyncWrite> AsyncRead for MaybeTlsStream<S> {}
+
+impl<S: Read + Write> Write for MaybeTlsStream<S> {
```
Comment (bruceg) [inline src/tls.rs]: Do these impl constraints have to be `Read + Write` or can they be just one or the other (ie `Read` for the `impl Read` etc)?

File: Cargo.toml [hunk 3]
```diff
@@ -138,7 +138,7 @@ inventory = "0.1"
 maxminddb = { version = "0.13.0", optional = true }
 strip-ansi-escapes = { version = "0.1.0", optional = true }
 colored = "1.9"
-warp = { version = "0.1.20", optional = true }
+warp = { version = "0.1", git = "https://github.com/timberio/warp", branch = "0.1.x", default-features = false, optional = true }
```
Comment (bruceg) [inline Cargo.toml]: Does this need documenting why we're on a moving target git branch here?

File: src/tls.rs (hunk 2 above)
Comment (LucioFranco) [reply src/tls.rs]: Unfortunately, it seems all of them require both bounds to get the inner tls stream to implement the correct one.

File: Cargo.toml (hunk 3 above)
Comment (LucioFranco) [reply Cargo.toml]: I can add something for that! Good idea!

**Comment by LucioFranco** [comment]
I am now running into feature flag issues where I am unable to figure out how to enable something for all the sources? @a-rodin any ideas how we should handle that?

**Comment by binarylogic** [comment]
@LucioFranco can you share more info about that actual problem? And what have you tried any workarounds?

**Comment by LucioFranco** [comment]
@binarylogic it has to do with how we est feature flags and their combinations. We don't really have a feature flag that gets enabled for all sources for things like `tls`.

@a-rodin I think this work around should work where I created a `sources-tls` feature flag, though this worries me since I feel like we will end up adding a bunch more and increasing the feature flag permutation check but a lot.

Another thing to note is I know we added this to reduce the time it takes to test but running the feature flag check takes about 10min on my machine which is decently long.

**Comment by bruceg** [comment]
I share the concerns about the explosion of feature flags. I understand the reasoning, but it is really adding complications to the build test process.

**Comment by binarylogic** [comment]
First, I would love to hear solutions, even if it's just removing all of the flags.

Second, if a user wanted to create a custom build with select components, how would they do that without feature flags? This is not a hard requirement, but it is a nice option for certain environments, such as embedded environments. Again, I could be wrong in my assumptions here and it is not a current requirement.

**Comment by LucioFranco** [comment]
@binarylogic I think I found a solution for now that works with the flags but I am cautious because I spent about an hour trying to get it all to compile since the feedback loop for checking all the flags is about 10min per full check run.

> Second, if a user wanted to create a custom build with select components, how would they do that without feature flags? 

Do we have any user asking for this or anywhere that explains what constraints they have? Basically, besides things like `leveldb`/`openssl`/`journald` everything is mostly in pure rust so the only issue with compiling more components is binary size which is already very small. I originally added feature flags around our more crazy deps like openssl and leveldb since those were mainly the ones that caused issues for me in non normal environments. Beyond that I don't see terrible benefit in say not compiling the tcp source or splunk sink. Since, both of those are pure rust and besides tls have no real issue cross compiling etc.

**Comment by LucioFranco** [comment]
Merging to unblock @bruceg

**Pull request merged**